pkg/operation/botanist/etcdencryption.go

```diff
@@ -0,0 +1,321 @@
+// limitations under the License.
+
+package botanist
+
+import (
+	"context"
+	"fmt"
+
+	gardenv1beta1 "github.com/gardener/gardener/pkg/apis/garden/v1beta1"
+	"github.com/gardener/gardener/pkg/operation/common"
+	encryptionconfiguration "github.com/gardener/gardener/pkg/operation/etcdencryption"
+	"github.com/gardener/gardener/pkg/utils"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiserverconfigv1 "k8s.io/apiserver/pkg/apis/config/v1"
+	"k8s.io/client-go/util/retry"
+)
+
+// CreateEtcdEncryptionConfiguration creates a secret
+func (b *Botanist) CreateEtcdEncryptionConfiguration(ctx context.Context) error {
+	enableEtcdEncryption, err := utils.CheckVersionMeetsConstraint(b.Shoot.Info.Spec.Kubernetes.Version, ">= 1.13")
+	if err != nil {
+		return err
+	}
+	if !enableEtcdEncryption {
+		return nil
+	}
+	needToWriteConfig := false
+	exists, ecSeed, forcePlaintextSecrets, err := b.readEncryptionConfigurationFromSeed()
+	if err != nil {
+		return err
+	}
+	if !exists {
+		// create new passive EncryptionConfiguration if it does not exist yet
+		// and remember to write this created configuration
+		b.Logger.Info("no EncryptionConfiguration found in seed. Creating new passive configuration.")
+		ecSeed, err = encryptionconfiguration.CreateNewPassiveConfiguration()
+		if err != nil {
+			return err
+		}
+		needToWriteConfig = true
+	} else {
+		exists, ecGarden, err := b.readEncryptionConfigurationFromGarden()
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return fmt.Errorf("EncryptionConfiguration exists in seed but not in garden")
+		}
+		// if it exists already, it needs to be consistent
+		consistent, err := b.isEncryptionConfigurationConsistent(ecSeed, ecGarden)
+		if err != nil {
+			return err
+		}
+		if !consistent {
+			return fmt.Errorf("EncryptionConfiguration in seed is not consistent")
+		}
+		if forcePlaintextSecrets {
+			if encryptionconfiguration.IsActive(ecSeed) {
+				// if annotation requires plaintext secrets, deactivate configuration if it is still active
+				b.Logger.Infof("EncryptionConfiguration is annotated with %v. De-activating existing active EncryptionConfiguration. Note: shoot secrets will be decrypted in etcd subsequently", common.EtcdEncryptionForcePlaintextAnnotationName)
+				encryptionconfiguration.SetActive(ecSeed, false)
+				needToWriteConfig = true
+			}
+		} else if !encryptionconfiguration.IsActive(ecSeed) {
+			// if it is not active already (meaning that identity and not aescbc
+			// is still the first encryption provider) then set it to active
+			// and remember to write this created configuration
+			b.Logger.Info("activating existing passive EncryptionConfiguration")
+			encryptionconfiguration.SetActive(ecSeed, true)
+			needToWriteConfig = true
+		}
+	}
+	ecSeedYamlBytes, err := encryptionconfiguration.ToYAML(ecSeed)
+	if err != nil {
+		return err
+	}
+	if needToWriteConfig {
```

Comment:
Maybe I miss something, but as far as I can see `needToWriteConfig` is always `true` if we reach this point, correct? Then we can remove it completely.

Reply:
No, it is only true if the configuration has changed in this reconciliation run (e.g. was initially created passive, was made "active", was moved back to passive due to the corresponding annotation).

Reply:
I checked again. If this part of the code is reached then it is always `true`, so there is no need to have this variable at all?

Reply:
lets have a look together ... I still disagree with you :) ...

Reply:
Okay, I see it now. As you can see it's hardly readable. Can you please refactor this part of the code as follows: Move the computation of the checksum to the top (as you always want to this). Then remove the `needToWriteConfig` variable and revert the control flow, i.e., exit early if you identify that the config is not needed to be rewritten.

Reply:
Hmm, in another of the 170 comments you were unhappy with me calling the writing function several times ... which is why I introduced the needToWriteConfig to remember whether I need to write or not ... now you are unhappy with that too ...

Reply:
OK, well, the problem is that this function is hardly readable for me (see comment history).
I suggest making it easier and less complex.
What about the following:

```golang
func (b *Botanist) CreateEtcdEncryptionConfiguration(ctx context.Context) error {
	ecSeed, forcePlaintextSecrets, err := b.readEncryptionConfigurationFromSeed()
	if err != nil {
		if !apierrors.IsNotFound(err) {
			return err
		}

		b.Logger.Info("no EncryptionConfiguration found in seed. Creating new passive configuration.")
		ecSeed, err = encryptionconfiguration.NewPassiveConfiguration()
		if err != nil {
			return err
		}

		return b.updateEtcdEncryptionConfiguration(ecSeed, forcePlaintextSecrets, true)
	}

	// The etcd encryption configuration does already exist, let's first check whether it is consistent with
	// the one that was replicated to the project namespace in the garden cluster.
	ecGarden, err := b.readEncryptionConfigurationFromGarden()
	if err != nil {
		if apierrors.IsNotFound(err) {
			// TODO: See question below, please.
			return fmt.Errorf("EncryptionConfiguration exists in seed but not in garden")
		}
		return err
	}

	if err := b.isEncryptionConfigurationConsistent(ecSeed, ecGarden); err != nil {
		return err
	}

	// The etcd encryption config is consistent. If it is already active and we are now forced to write all
	// secrets in plain text then let's make it inactive.
	if encryptionconfiguration.IsActive(ecSeed) && forcePlaintextSecrets {
		b.Logger.Infof("EncryptionConfiguration is annotated with %v. De-activating existing active EncryptionConfiguration. Note: shoot secrets will be decrypted in etcd subsequently", common.EtcdEncryptionForcePlaintextAnnotationName)
		encryptionconfiguration.SetActive(ecSeed, false)
		return b.updateEtcdEncryptionConfiguration(ecSeed, forcePlaintextSecrets, true)
	}

	// The etcd encryption config is consistent but not yet active. Let's activate it now.
	if !encryptionconfiguration.IsActive(ecSeed) {
		b.Logger.Info("activating existing passive EncryptionConfiguration")
		encryptionconfiguration.SetActive(ecSeed, true)
		return b.updateEtcdEncryptionConfiguration(ecSeed, forcePlaintextSecrets, true)
	}

	// The etcd encryption config is consistent + active, but we are not forced to write the secrets in plain text.
	// We don't need to rewrite the configuration but will compute the checksum for later usage during kube-apiserver
	// configuration.
	return b.updateEtcdEncryptionConfiguration(ecSeed, forcePlaintextSecrets, false)
}

func (b *Botanist) updateEtcdEncryptionConfiguration(ecSeed *TODO, forcePlaintextSecrets, needToWriteConfig bool) error {
	ecSeedYamlBytes, err := encryptionconfiguration.ToYAML(ecSeed)
	if err != nil {
		return err
	}

	if needToWriteConfig {
		if err := b.writeEncryptionConfiguration(ecSeedYamlBytes, forcePlaintextSecrets); err != nil {
			return err
		}
	}

	b.mutex.Lock()
	b.CheckSums[common.EtcdEncryptionSecretName] = utils.ComputeSHA256Hex(ecSeedYamlBytes)
	b.mutex.Unlock()

	return nil
}
```

Please note that I removed the `exists` return parameter from the two functions - it's a better practice to check the error via the apimachinery utilities (also improves readability). You might consider doing the same for other functions (if any) as well.

Also, see the `TODO` in the code above. Why do we actually error in case we have not found the etcd encryption config in the garden cluster? Let's assume we create the configuration and are able to write it to the seed, but we cannot write it to the garden cluster because of network problems (or whatever). In the next try we will throw this error which will never automatically resolve itself --> Requiring ops effort and manual intervention.
Should we only do the consistency check if the config actually exists in the garden cluster?

Reply:
okay, I see that this improves the readability of the code. Will incorporate the change. 

With regards to the TODO question ... it is unclear to me how much magic to implement to cover error cases. It would be nice if I could bundle all the changes in a sort of transaction together and only actually do the EncryptionConfiguration change at BOTH sides or not at all (rollback). 

The whole point of storing the configuration at two sides was to minimize the potential for loosing the EncryptionConfiguration. Therefore if I cannot read it from both locations and the result is equal I chose to fail.

But you are correct, if I wrote once to the seed, and the Garden write fails, then I cannot get out of this without manual intervention. 

Would you propose to ignore this and just hope to write and ignore if a write fails? Or even try to remove the configuration from the seed again if the write to Garden fails? I actually do not want to make these decisions and thereby risk loosing an in-use Configuration ...

Reply:
>  It would be nice if I could bundle all the changes in a sort of transaction together and only actually do the EncryptionConfiguration change at BOTH sides or not at all (rollback).

That's not possible easily.

> Would you propose to ignore this and just hope to write and ignore if a write fails?

I would propose to ignore it and dump the etcd encryption config into the log in this case.

> Or even try to remove the configuration from the seed again if the write to Garden fails?

Well, if you want to make sure that the information exists in the garden clusters as well then why not write first to the garden cluster, and only if this succeeded, write it to the seed and "activate" it?

Reply:
The SAP Security Standard dictates, that no password/secrets/keys should appear in any logs, therefore I would rather adhere to the standard here :) ... 

I am unhappy with ignoring this error ... since doing so would mean that the key is stored only at one location (the seed) and doing the automatic rewriting the secrets based on that seems to me not such a good idea. I would rather have the task fail and subsequently also the automatic secret rewrite task fail. 

But clearly this might happen too if the write after the activation of the EncryptionConfiguration fails and then the configuration in the seed is active ... Although the automatic rewriting of secrets does not occur, any secrets rewritten by other tasks would be encrypted already. Nothing to be done about that I am afraid ... At least risk is minimized ... wouldn't you agree?

Reply:
Yes, but what about my proposal "reverting" the order? If you deploy the configuration FIRST to the garden cluster and ONLY if this succeeded you deploy it to the seed (both happens in the first task `CreateEtcdEncryptionConfiguration`) then it should be fine, correct?

If the write operation to the garden cluster fails you don't continue. This means that the kube-apiserver won't be deployed and the flow aborts.
If the write operation to the garden cluster succeeds but the write operation to the seed cluster fails you abort as well. Only if both write operations succeeded you allow continuing with deployment of kube-apiserver (which might activate the configuration)?

Reply:
As you suggested, I reorder and now first write to garden and then to seed. If the write to the seed fails, I replace the updated version just written to garden with the old prior version.